### PR TITLE
Remove weight gradients if model is frozen for linear layer

### PR DESCRIPTION
**Pull request by JimmyZhang12**
Changes the backward pass of the linear layer so that weight gradients for frozen models are not calculated

* Checks if weights have `requires_grad`
* if `requires_grad=False`, only calculate and return the data gradients
* Improves training step time for prompt tuning gpt3 by about 20%

### CODE REVIEW

File: apex/transformer/tensor_parallel/layers.py [hunk 1]
```diff
@@ -313,45 +320,76 @@ def forward(
...
+                    sub_grad_input,
+                    grad_input,
+                    group=get_tensor_model_parallel_group(),
+                    async_op=True
+                )
+                handle.wait()
+                return sub_grad_input, None, None, None, None, None, None
+            if ctx.async_grad_allreduce:
+                handle.wait()
+            return grad_input, None, None, None, None, None, None
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/layers.py:385]: QQ: is this `grad_input` dependent on L.355 of `grad_input = grad_output.matmul(weight)` and L.362 when async_grad_allreduce? (I'm trying to understand why the view op on grad_output and total_input are moved

File: apex/transformer/tensor_parallel/layers.py [hunk 2]
```diff
@@ -283,12 +283,19 @@ def forward(
         sequence_parallel_enabled: bool,
         use_16bit_in_wgrad_accum_fusion: bool = False,
     ):
-        ctx.save_for_backward(input, weight)
         ctx.use_bias = bias is not None
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/layers.py:286]: nit-picking

```suggestion
        ctx.use_bias = bias is not None and weight.requires_grad
```

File: apex/transformer/tensor_parallel/layers.py [hunk 3]
```diff
@@ -283,12 +283,19 @@ def forward(
...
         ctx.async_grad_allreduce = async_grad_allreduce
         ctx.sequence_parallel_enabled = sequence_parallel_enabled
         ctx.use_16bit_in_wgrad_accum_fusion = use_16bit_in_wgrad_accum_fusion
+        ctx.compute_weight_gradient = weight.requires_grad
+
+        if ctx.compute_weight_gradient:
+            ctx.save_for_backward(input, weight)
+        else:
+            ctx.save_for_backward(weight)
+            ctx.use_bias = False
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/layers.py:297]:
```suggestion
```

File: apex/transformer/tensor_parallel/layers.py [hunk 4]
```diff
@@ -313,45 +320,76 @@ def forward(
...
-                group=get_tensor_model_parallel_group(),
-                async_op=True,
-            )
-            total_input = all_gather_buffer
-        else:
-            total_input = input
         grad_input = grad_output.matmul(weight)
 
-        if ctx.sequence_parallel_enabled:
+        if handle is not None:
```
Comment (crcrpar) [inline apex/transformer/tensor_parallel/layers.py:357]: Is this condition equivalent to `ctx.sequence_parallel_enabled and ctx.compute_weight_gradient`?

File: apex/transformer/tensor_parallel/layers.py (hunk 1 above)
Comment (JimmyZhang12) [reply apex/transformer/tensor_parallel/layers.py:385]: The view on grad_output and total input are only used for calculating the weight gradient so if `weight.requires_grad=False` then they are not needed so I moved them after the return on L.385 so that views would not be executed. Maybe it would be better to keep the views as they are but wrap them in an `if ctx.compute_weight_grad`

File: apex/transformer/tensor_parallel/layers.py (hunk 4 above)
Comment (JimmyZhang12) [reply apex/transformer/tensor_parallel/layers.py:357]: Yes it is equivalent

File: apex/transformer/tensor_parallel/layers.py (hunk 1 above)
Comment (crcrpar) [reply apex/transformer/tensor_parallel/layers.py:384]: understood. when I first saw "immediately return", I wondered if we could put the no weight grads logic before with weight grads.

**Review by crcrpar — APPROVED** [review]

**Comment by crcrpar** [comment]
@JimmyZhang12 could you resolve conflicts, following https://github.com/NVIDIA/apex/pull/1513?

**Comment by JimmyZhang12** [comment]
Conflicts should be resolved now.

**Review by crcrpar — APPROVED** [review]
